test/e2e/clusterctl_upgrade.go

```diff
@@ -133,6 +135,15 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 		} else {
 			clusterctlBinaryURLTemplate = input.InitWithBinary
 		}
+
+		// create InitWithKubernetesVersion field to set the init kubernetes version via struct
+		if input.InitWithKubernetesVersion == "" {
+			Expect(input.E2EConfig.Variables).To(HaveKey(initWithKubernetesVersion), "Invalid argument. %s variable must be defined when calling %s spec", initWithKubernetesVersion, specName)
+			Expect(input.E2EConfig.Variables[initWithKubernetesVersion]).ToNot(BeEmpty(), "Invalid argument. %s variable can't be empty when calling %s spec", initWithKubernetesVersion, specName)
+			input.InitWithKubernetesVersion = input.E2EConfig.GetVariable(initWithKubernetesVersion)
+		} else {
+			input.InitWithKubernetesVersion = os.Getenv(initWithKubernetesVersion)
```

Comment:
I don't think we need the else branch. input.E2EConfig.GetVariable does this internally

Reply:
Overall the logic should be 
* if input.InitWithKubernetesVersion is set, don't modify it
* if input.InitWithKubernetesVersion is not set get it from variables (which looks for env variables and entries in the e2e config)